databricks/sdk/oauth.py

```diff
@@ -187,21 +189,131 @@ def retrieve_token(client_id,
+    def _async_token(self) -> Token:
+        """
+        Returns a token.
+        If the token is stale, triggers an asynchronous refresh.
+        If the token is expired, refreshes it synchronously, blocking until the refresh is complete.
+        """
+        token_state = self._token_state()
+
+        with self._lock:
+            token = self._token
+
+        if token_state == _TokenState.FRESH:
+            return token
+        if token_state == _TokenState.STALE:
+            self._trigger_async_refresh()
+            return token
+        return self._blocking_token()
+
+    def _token_state(self) -> _TokenState:
+        """Returns the current state of the token."""
+        with self._lock:
+            if not self._token or not self._token.valid:
+                return _TokenState.EXPIRED
+            if not self._token.expiry:
+                return _TokenState.FRESH
+
+            lifespan = self._token.expiry - datetime.now()
+            if lifespan < timedelta(seconds=0):
+                return _TokenState.EXPIRED
+            if lifespan < self._stale_duration:
+                return _TokenState.STALE
+            return _TokenState.FRESH
+
+    def _blocking_token(self) -> Token:
+        """Returns a token, blocking if necessary to refresh it."""
+        state = self._token_state()
+        with self._lock:
+            # This is important to recover from potential previous failed attempts
+            # to refresh the token asynchronously.
+            self._refresh_err = False
+            self._is_refreshing = False
+
+            # It's possible that the token got refreshed (either by a _blocking_refresh or
+            # an _async_refresh call) while this particular call was waiting to acquire
+            # the lock. This check avoids refreshing the token again in such cases.
+            if state != _TokenState.EXPIRED:
                 return self._token
+
             self._token = self.refresh()
             return self._token
-        finally:
-            self._lock.release()
+
+    def _trigger_async_refresh(self):
+        """Starts an asynchronous refresh if none is in progress."""
+
+        def _refresh_internal():
+            new_token: Token = None
+            try:
+                new_token = self.refresh()
+            except Exception as e:
+                # This happens on a thread, so we don't want to propagate the error.
+                # Instead, if there is no new_token for any reason, we will disable async refresh below
+                # But we will do it inside the lock.
+                logger.warning(f'Tried to refresh token asynchronously, but failed: {e}')
+
+            with self._lock:
+                if new_token is not None:
+                    self._token = new_token
+                else:
+                    self._refresh_err = True
+                self._is_refreshing = False
+
+        with self._lock:
+            if not self._is_refreshing and not self._refresh_err:
```

Comment:
I think this is safe but might fetch more tokens than necessary. Like you mentioned offline, the first thread to acquire this lock will start the refresh process, and then all other threads that acquire the lock while refresh is happening will skip. However, once refresh is complete, other threads that acquire the lock (e.g. that checked that the token was stale and came down this pathway while the refresh was ongoing, then only acquired this lock after refresh was completed) will unnecessarily refresh the token again. So I think we should also check the token state again here: if it is not stale, skip the refresh.